### PR TITLE
Default CanvasItem materials to use sRGB space for uniform colors

### LINKED ISSUE #61697: [4.0a9] inconsistent sRGB in canvas_item shaders
### Godot version

4.0 alpha 9 (d9daf3869)

### System information

Windows 11, GLES3, NVIDIA GeForce RTX 3050

### Issue description

Comparing a uniform to COLOR in 4.0 gives inconsistent results for unknown reasons. Likely having to do with sRGB/linear conversion.

### Steps to reproduce

Provided is a working (non-bugged) minimal reproduction in v3.5, and a non-working (bugged) minimal reproduction in v4.0.

Open srgb_3.5 in Godot v3.5-rc1. Open srgb_4.0a9 in Godot v4.0 alpha 9. Note the difference in color: srgb_3.5 has colored the correct area, and srgb_4.0a9 has colored the wrong area.

Then in srgb_4.0a9, click on Sprite->Inspector->Material->Shader->Shader Param, and set the sensitivity to 0.01, which is the same as srgb_3.5's sensitivity. Note how the shader now does nothing.

### Minimal reproduction project

[srgb_3.5.zip](https://github.com/godotengine/godot/files/8838407/srgb_3.5.zip)
[srgb_4.0a9.zip](https://github.com/godotengine/godot/files/8838416/srgb_4.0a9.zip)

### LINKED ISSUE #66572: Shader using source_color uses a more saturated color than the actual one chosen
### Godot version

v4.0.beta1.official [20d667284]

### System information

Arch Linux, Vulkan, Intel HD Graphics 4400

### Issue description

The shader uses a more saturated version of the color you give it
image

The shader parameter for `color` is lighter, and yet shows up as a more saturated color on the preview and the object itself.

This is my first time making an issue on this GitHub, So I may have done something incorrectly, If I have please let me know!

### Steps to reproduce

You can write a shader as simple as this:
```
shader_type canvas_item;

uniform vec4 color : source_color;

void fragment() {
    COLOR = color;
}
```

And use it on an object (I've only tested with a Sprite2D, i don't know if this happens on 3D objects)
Choose a lighter color for the `color` shader parameter and watch it be more saturated.

I have uploaded the ZIP file for a reproduction project. It includes an example scene with a sprite that has the shader using the code above.

### Minimal reproduction project

[Shader Issue.zip](https://github.com/godotengine/godot/files/9669091/Shader.Issue.zip)

### PR DESCRIPTION
**Pull request by clayjohn**
Fixes: https://github.com/godotengine/godot/issues/61697
Fixes: https://github.com/godotengine/godot/issues/66572

The 2D renderer in Godot is totally in sRGB space so it is appropriate to keep 2D uniform colors in sRGB space. Specifying uniform colors in the color picker should now match the actual rendered color in both the 2D and 3D renderers.

### History

There is a bit of history here that warrants mention. The original behaviour was introduced in https://github.com/godotengine/godot/commit/c43f624d44fe143364154ee232332e17a4fddc3d In that PR, the ``update_uniform_buffer()`` function was added with behaviour that depended on the ``p_use_linear_color`` flag (if true, converts colors to linear before passing to shader). It  was always ``false`` so all uniforms were passed to shaders in sRGB space. It appears to me that @reduz meant to make this value conditional on whether the shader was a 2D or 3D shader, but never finished.

That led to problems with 3D shaders that used uniform colors (like our editor line drawing) I "fixed the bug by defaulting to ``true`` which meant that anytime ``source_color`` was used, the value was converted to linear. This of course led to the issue we face now. All uniform colors are now linear despite the need for sRGB colors in canvas item shaders. 

### The solution

The solution I have implemented is to treat CanvasItem shaders different from the other shader types. I beleive this is the implementation @reduz had in mind when he made https://github.com/godotengine/godot/commit/c43f624d44fe143364154ee232332e17a4fddc3d as he exposed the parameter in the first place but never used it. 

### The problem with the solution
The problem with the solution is that it treats particles shaders as if they are always 3D. So users will still run into this issue or a similar one when authoring particles materials that use colors. 

Additionally, users have asked for a way to have a color picker popup for uniforms without doing sRGB->Linear conversions. 

### The alternative
Essentially, a possible API would be to separate sRGB/Linear conversion from whether their is a color picker. I.e. ``source_color, use_linear`` would expose a color picker and would convert to linear before passing (this would be used for 3D) while ``source_color`` would only expose the color picker (this would be used for 2D).

Essentially this PR (and the status quo) take the complexity away from the user. The shader essentially says "tell me that you are sending me a color and I will make sure it looks good". The alternative I have proposed above would push the sRGB/linear decision to the user who would be responsible for using the ``use_linear`` flag in spatial shaders. In my opinion, this is a valid design choice, but is counter to the "Godot-way" of doing things which is to expose minimal complexity and make things that "just work". 

In summary, in my opinion, we should stick with this PR instead of the alternative, but I want to have the reasoning fully set out. 

### WARNING
**Note: As a workaround users have been manually converting back to sRGB in shaders. This PR will appear to break those materials. Unfortunately, this is necessary breakage to fix the underlying issue.**

### CODE REVIEW

**Review by akien-mga — APPROVED** [review]
Makes sense to me.

I agree that doing the "most appropriate" conversion transparently is good.

That being said, we might want to add `use_linear` and `use_srgb` hints for situations where users need to force it - notably 2D particle shaders IIUC?

**Comment by akien-mga** [comment]
Thanks!

**Comment by clayjohn** [comment]
> Makes sense to me.
> 
> I agree that doing the "most appropriate" conversion transparently is good.
> 
> That being said, we might want to add `use_linear` and `use_srgb` hints for situations where users need to force it - notably 2D particle shaders IIUC?

Ya we could, users may also be fine just doing the conversion to sRGB manually (as many are doing in canvas_item shaders now)
